### ISSUE TITLE
netCDF configuration

### ISSUE THREAD

**Issue by johnomotani** [boutproject/BOUT-dev#1898]
On Marconi, I have to explicitly pass the `lib` (in `EXTRA_LIBS`) and `include` (in `EXTRA_INCS`) search paths for `netcd-c` to `configure`*. I suspect the problem may be due to `configure` only using `ncxx4-config` provided by `netcdf-cxx4` to find the paths to add. It would be nice if `configure` would **also** get paths from `nc-config` (provided by `netcdf-c`). I guess this might need an extra config flag: maybe we could have both `with-netcdf` and `with-netcdf-cxx` or something like that, for systems which need both?

\* possibly part of the problem is that sometimes there is a very old `netcdf` provided by the system, which gets found instead of one provided by modules, or installed by the user, etc.

**Comment by dschwoerer** [comment]
Other question: Do you have a somewhat working setup for marconi? Currently I use everything compiled from source, except the gcc compiler ...
If so, would you mind sharing your setup in the documentation?

**Comment by johnomotani** [comment]
@dschwoerer I don't have a nice setup for Marconi. I'm using `gnu/7.3.0`, `openmpi/4.0.1--gnu--7.3.0`, and `hdf5/1.10.5--openmpi--4.0.1--gnu--7.3.0`, but compiling `netcdf`, `netcdf-cxx`, `sundials` and `fftw` from source. Support in the modules should be if you use intel compilers, but since we found that they tend to be slower I've not checked recently.

**Comment by ZedThree** [comment]
What versions of `netcdf-c` and `netcdf-cxx4` are you using? I'm reasonably certain that I fixed both of them several years ago.

Are you installing `netcdf-cxx4` into the same installation directory as `netcdf-c`? If not, I do recommend you do so, as various bits of their build system assume you have done.

I thought I'd also fixed the cxx4 cmake installation too, but it looks like my PR hasn't been merged :(

**Comment by johnomotani** [comment]
Ah installing in the same directory would get around the problem... I haven't been because I used to be using the module-provided `netcdf-c`, but having to compile `netcdf-cxx4` myself. Now I have to compile both myself anyway I could change that...

**Comment by johnomotani** [comment]
Installing/linking `netcdf-cxx4` has often been a bit of a pain. Can we switch to using the C interface at some point so we don't need `netcdf-cxx4` any more?

**Comment by johnomotani** [comment]
> Are you installing netcdf-cxx4 into the same installation directory as netcdf-c? If not, I do recommend you do so, as various bits of their build system assume you have done.

I think it would be nice to have some work-around for if they are not installed in the same directory, for the not-uncommon case when a cluster provides `netcdf-c` but not `netcdf-cxx4`. I think just adding the output from `nc-config --include` to `EXTRA_INCS` and `nc-config --libs` TO `EXTRA_LIBS` before the corresponding ones from `ncxx4-config` should be enough?

**Comment by Thrysoe** [comment]
I found this thread looking for a working setup for MARCONI for v. 4.3.1 and saw the same question in a previous post. Has the status on this changed since asked in January?

**Comment by johnomotani** [comment]
No this hasn't changed. I actually ended up switching to compiling netcdf-c for myself as well, which actually simplified the BOUT++ configuration a bit because I could just install netcdf-c and netcdf-cxx4 into the same directory.

Some of the difficulty is that (at least last time we at Culham checked) gnu compilers produce faster BOUT++ code than Intel ones, but Marconi's modules are mostly set up around Intel. They provide some libraries for using with gcc, but only for some versions of gcc (often an older one).

So I've currently got `gnu/7.3.0`, `openmpi/4.0.1--gnu--7.3.0` and `hdf5/1.10.5--openmpi--4.0.1--gnu--7.3.0` loaded from modules (which I think needed `profile/advanced`, `szip/2.1--gnu--6.1.0` and `zlib/1.2.8--gnu--6.1.0` and possibly some of the Intel compilers/libraries as prerequisites). Then I downloaded source code for `fftw-3.3.8`, `netcdf-c-4.7.4`, `netcdf-cxx4-4.3.1`, and compiled in our $WORK directory. I'm also using `sundials-5.1.0` and `petsc-3.13.0`, again compiled from source. Then I think the only 'trick' is to use the same `--prefix` argument when configuring `netcdf-c` and `netcdf-cxx4` so that they install into the same directory.

If you have trouble, I suggest we discuss on Slack until you get something working, then maybe you could post the final solution on here to help others in future...

**Comment by Thrysoe** [comment]
Thank you so much! This worked for me, with only minor adjustments due to available compilers and an issue with netcdf-c-4-7.4 which was suggested. The latter gave an error that that HDF5 library was not built with zlib so I used a previous release which worked just fine. The modules (and profiles) that I load on MARCONI are:

 1) profile/base       3) gnu/7.3.0                   5) zlib/1.2.8--gnu--6.1.0   7) hdf5/1.8.20--openmpi--3.0.0--gnu--7.3.0  
 2) profile/advanced   4) openmpi/3.0.0--gnu--7.3.0   6) szip/2.1--gnu--6.1.0 

From source I installed: `fftw-3.3.8`, `netcdf-c-4.6.3`, and `netcdf-cxx4-4.3.1` in the same folder as suggested. It also got an error when removing -lnetcdf from the EXTRA_LIBS flag in make.config, which used to be necessary.
